### PR TITLE
Add flag data format

### PR DESCRIPTION
**Pull request by LucasMahieu**
* Add a parameter for all handler front-end functions to give the previous nodes to the current node in the graph in order to be able to get the input shape tensor for examples
* Add a flag that user can specify in order to know when converting to/from TF which data format it is used. Then, it is possible to add transpose of to change parameters (axis, perm, shape, etc)
* Add a flag that user can specify in order to know when converting to/from TF for which device type the graph should be compatible : CPU or CUDA.

### CODE REVIEW

File: test/backend/test_model.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def test_relu_node_inplace(self):
       inputs=[helper.make_tensor_value_info("X", TensorProto.FLOAT, [3, 2])],
       outputs=[helper.make_tensor_value_info("X1", TensorProto.FLOAT, [3, 2])])
     tf_rep = prepare(helper.make_model(graph_def))
+
```
Comment (tjingrant) [inline test/backend/test_model.py]: These changes seem irrelevant, would you mind cleaning up a bit so the commit history will not be unnecessarily long?

File: test/frontend/test_node.py [hunk 2]
```diff
@@ -38,13 +38,15 @@ class TestNode(unittest.TestCase):
...
-  test_option = test_data[5] if len(test_data) > 5 else {}
+  test_option = test_data[7] if len(test_data) > 7 else {}
 
   def do_test_expected(self):
     tf_op = test_data[1]
     output_name = test_data[2]
     inputs = test_data[3]
     attrs = test_data[4]
+    device = test_data[5]
+    channel_last = test_data[6]
```
Comment (tjingrant) [inline test/frontend/test_node.py]: Rather than specifying these two attributes, can we enumerate through all possibilities? For instance, if there's a GPU, then test NHWC and NCHW, if not, only test NHWC. Checks can be done via `supports_device` function.

This won't have any effects on CI, but I can test them locally on my work machine with GPU. This could significantly increase confidence in correctness.

File: onnx_tf/frontend.py [hunk 3]
```diff
@@ -217,14 +264,23 @@ def handle_random_standard_normal(cls, node, consts):
...
         and offsetted random standard uniform op.
     """
+    shape = node.attr["_output_shapes"][0]
+    # Arrange shape to the according data format 
+    if cls._channel_last:
+      rank = len(shape)
+      if rank >= 3:
+        old_shape = shape[:]
+        for i, s in enumerate(shape):
+          j = i % (rank-1) + 1 if i != 0 else 0
```
Comment (tjingrant) [inline onnx_tf/frontend.py]: I have seen this pattern quite a bit. Can you use a helper function to perform this NHWC->NCHW transformation? This could increase readability quite a bit.

File: onnx_tf/frontend.py [hunk 4]
```diff
@@ -152,20 +168,25 @@ def tensorflow_graph_to_onnx_graph(cls, graph_def, output, name="graph"):
 
         # Check if specialized handler exists.
         if handler_name in dir(cls):
+          input_node = get_nodes_by_name(graph_def.node, node.inputs)
           method_to_call = getattr(cls, handler_name)
-          ops_proto.append(method_to_call(node, consts))
+          ops_proto.append(method_to_call(node, consts, input_node))
```
Comment (tjingrant) [inline onnx_tf/frontend.py]: The whole point of having a input_node param seems to be that you want to access the shapes of input nodes, which might be better suited as an attribute in the TensorflowNode class? What do you think? This way we can keep the signature of handler clean (important because this signature is copied in every handler...) as most handlers do not need this param.

**Comment by tjingrant** [comment]
Hi, thanks a lot for your hard work, I've left a few small suggestions here and there; 

One important thing is that tests on GPU are important b/c this is what most people do anyways.

File: onnx_tf/frontend.py (hunk 4 above)
Comment (LucasMahieu) [reply onnx_tf/frontend.py]: Yes, I agree with you for the signature. It should be betted to keep it without the input node param.
But I did it because I didn't get the idea to add the input shape to the TensorflowNode class.

I will make the fix, thanks

File: onnx_tf/frontend.py (hunk 3 above)
Comment (LucasMahieu) [reply onnx_tf/frontend.py]: Of course, I will do the helper function as soon as possible.

File: test/frontend/test_node.py (hunk 2 above)
Comment (LucasMahieu) [reply test/frontend/test_node.py]: Ok, but with GPU, I think the only correct data format is NCHW, right ? 

Testing all possible data format more dynamically is a good idea.

Comment (tjingrant) [reply test/frontend/test_node.py]: To quote from this thread https://github.com/tensorflow/tensorflow/issues/2660:

> MaxPool on GPU supports NCHW and NHWC
> MaxPool on CPU only supports NHWC.

I'd expect the same for most other ops that care about input formats.

Comment (LucasMahieu) [reply test/frontend/test_node.py]: Can you try on a gpu to be really sur of that ?
Because it could impact my backend modifications in order to manage the data format...

**Comment by LucasMahieu** [comment]
Thanks a lot for your feed back, I will take into account all of them, then make a pull request again

File: test/frontend/test_node.py (hunk 2 above)
Comment (tjingrant) [reply test/frontend/test_node.py]: I'll let you know my trials later today but that statement was made by TF developers from Google so I'd consider that more authoritative than even my experimental results.

**Comment by tjingrant** [comment]
Hi, there's no need to make another pull request, you can simply continue committing to this branch and the commits will automatically show up here.

**Comment by LucasMahieu** [comment]
Ho, great !

**Comment by LucasMahieu** [comment]
Here is my update, do not hesitate to comment.

File: test/frontend/test_node.py [hunk 5]
```diff
@@ -94,24 +116,44 @@ def do_test_expected(self):
 ("test_or", tf.logical_or, "LogicalOr", [get_rnd([10, 10], dtype=np.bool_), get_rnd([10, 10], dtype=np.bool_)], {}),
 ("test_pow", tf.pow, "Pow", [get_rnd([10, 10]), get_rnd([10, 10])], {}),
 ("test_pad", tf.pad, "Pad", [get_rnd([2, 3]), [[1, 1,], [2, 2]]], {"mode": "constant"}),
+("test_pad", tf.pad, "Pad", [get_rnd([10, 2, 2, 3]), [[0,0], [2, 2], [2, 2], [0,0]]], {"mode": "constant"}),
```
Comment (tjingrant) [inline test/frontend/test_node.py]: Test name has to be unique. Otherwise only one of these gets run.

**Comment by tjingrant** [comment]
Looking good, let's merge as soon as I run through all the tests on a GPU!

File: test/frontend/test_node.py (hunk 5 above)
Comment (LucasMahieu) [reply test/frontend/test_node.py]: Oh, thanks to notice it to me.
Do you want to make to change before to merge ? 
Or do you want I make the modification ?

Comment (tjingrant) [reply test/frontend/test_node.py]: Could you modify them to have unique names?

Comment (LucasMahieu) [reply test/frontend/test_node.py]: I can do it ASAP.

**Comment by LucasMahieu** [comment]
Here is my update, do not hesitate to comment.
I have made some modifications due to bugs discovered thanks to tests that was not executed.

**Comment by tjingrant** [comment]
Fantastic, I hope you find these troubles of working on tests worth going through!

Also, given that nothing has really changed (unfortunately) in the spec repo regarding the data format issues we face, do you mind if I edit your comment here https://github.com/onnx/onnx-tensorflow/issues/31#issuecomment-352019711 and add it to somewhere in the README? This is important information that the user needs to know but is currently undocumented.

**Comment by tjingrant** [comment]
Hi, I tried running tests on my GPU but GraphDef reached 2GB at some point and TF refuses to execute the graph. I think I need to add a tear_down function to the unit test in ONNX spec repo that resets the graph at some point...

**Comment by LucasMahieu** [comment]
Great, unit test is fundamental when developing software like this.
I also really think that it is necessary to warn the user in one way or another of the choice of implementation that I made.

For GPU problem I can't help you because I don't have any GPU ...
Do you think we can have access to GPU with CI travis check during pull request ?

**Comment by tjingrant** [comment]
OK, managed to test almost everything on GPU.
